packages/did-auth-siop-op-authenticator/src/agent/DidAuthSiopOpAuthenticator.ts

```diff
@@ -0,0 +1,166 @@
+import { schema } from '../index'
+import { IAgentPlugin } from '@veramo/core'
+import { OperatingPartySession } from '../session/OperatingPartySession';
+import {
+  ParsedAuthenticationRequestURI,
+  VerifiedAuthenticationRequestWithJWT,
+} from '@sphereon/did-auth-siop/dist/main/types/SIOP.types'
+import {
+  events,
+  IAuthenticateWithDidSiopArgs,
+  IAuthRequestDetails,
+  ICreateDidSiopSessionArgs,
+  IDidAuthSiopOpAuthenticator,
+  IGetDidSiopAuthenticationRequestDetailsArgs,
+  IGetDidSiopAuthenticationRequestFromRpArgs,
+  IGetDidSiopSessionArgs,
+  IRegisterCustomApprovalForDidSiopArgs,
+  IRemoveCustomApprovalForDidSiopArgs,
+  IRemoveDidSiopSessionArgs,
+  IRequiredContext,
+  ISendDidSiopAuthenticationResponseArgs,
+  IVerifyDidSiopAuthenticationRequestUriArgs
+} from '../types/IDidAuthSiopOpAuthenticator'
+
+/**
+ * {@inheritDoc IDidAuthSiopOpAuthenticator}
+ */
+export class DidAuthSiopOpAuthenticator implements IAgentPlugin {
+  readonly schema = schema.IDidAuthSiopOpAuthenticator
+  readonly methods: IDidAuthSiopOpAuthenticator = {
+    getDidSiopSession: this.getDidSiopSession.bind(this),
+    addDidSiopSession: this.addSessionForDidSiop.bind(this),
+    removeDidSiopSession: this.removeSessionForDidSiop.bind(this),
+    authenticateWithDidSiop: this.authenticateWithDidSiop.bind(this),
+    getDidSiopAuthenticationRequestFromRP: this.getDidSiopAuthenticationRequestFromRP.bind(this),
+    getDidSiopAuthenticationRequestDetails: this.getDidSiopAuthenticationRequestDetails.bind(this),
+    verifyDidSiopAuthenticationRequestURI: this.verifyDidSiopAuthenticationRequestURI.bind(this),
+    sendDidSiopAuthenticationResponse: this.sendDidSiopAuthenticationResponse.bind(this),
+    registerCustomApprovalForDidSiop: this.registerCustomApprovalForDidSiop.bind(this),
+    removeCustomApprovalForDidSiop: this.removeCustomApprovalForDidSiop.bind(this)
+  }
+
+  private readonly sessions: Record<string, OperatingPartySession>
+  private readonly customApprovals: Record<string, (verifiedAuthenticationRequest: VerifiedAuthenticationRequestWithJWT) => Promise<void>>
+
+  constructor(customApprovals?: Record<string, (verifiedAuthenticationRequest: VerifiedAuthenticationRequestWithJWT) => Promise<void>>) {
+    this.sessions = {}
+    this.customApprovals = customApprovals || {}
+  }
+
+  /** {@inheritDoc IDidAuthSiopOpAuthenticator.getSession} */
+  private async getDidSiopSession(
+      args: IGetDidSiopSessionArgs,
+      context: IRequiredContext
+  ): Promise<OperatingPartySession> {
+    if (this.sessions[args.sessionId] === undefined) {
+      return Promise.reject(new Error(`No session found for id: ${args.sessionId}`))
+    }
+
+    return this.sessions[args.sessionId]
+  }
+
+  /** {@inheritDoc IDidAuthSiopOpAuthenticator.addSessionForDidSiop} */
+  private async addSessionForDidSiop(
```

Comment:
I think we should allow for no sessionId to be passed in as well in which case we just generate a random sessionId. Probably makes sense to rename the method to 'new' instead of 'add' then as well

Reply:
sessionId is now optional and we will generate an uuid for if one is not provided

Reply:
Makes sense